### PR TITLE
fix: Validate embedded objects schema [DHIS2-2109]

### PR DESCRIPTION
**Pull request by enricocolasante**

### CODE REVIEW

**Comment by enricocolasante** [comment]
After talking with a few people, I restored old endpoints that were deleted and managed with the API versioning.

At the moment we have this versioned endpoint

`@ApiVersion( include = { DhisApiVersion.DEFAULT, DhisApiVersion.ALL }, exclude = { DhisApiVersion.V29, DhisApiVersion.V30, DhisApiVersion.V31, DhisApiVersion.V32, DhisApiVersion.V33 } )
public void getTaskJson( @PathVariable( "jobType" ) String jobType, @RequestParam( required = false ) String lastId, HttpServletResponse response )

@ApiVersion( include = { DhisApiVersion.V29, DhisApiVersion.V30, DhisApiVersion.V31, DhisApiVersion.V32, DhisApiVersion.V33 }, exclude = { DhisApiVersion.DEFAULT, DhisApiVersion.ALL } )
public void getTasksExtendedJson( @PathVariable( "jobType" ) String jobType, HttpServletResponse response ) throws IOException`

This seems wrong to me, this 2 methods are doing slightly different things and they should be 2 different methods.
So the way to proceed in order to avoid to break any clients is create a new `/extended/task/{jobType}` endpoint (that will be used by the frontend) and for the next releases add the version to the default task endpoint. In 3 versions we will remove the versioned endpoint.

File: dhis-2/dhis-services/dhis-service-dxf2/src/test/java/org/hisp/dhis/dxf2/metadata/objectbundle/ObjectBundleServiceTest.java [hunk 1]
```diff
@@ -33,11 +33,7 @@
 import org.hisp.dhis.category.Category;
 import org.hisp.dhis.category.CategoryCombo;
 import org.hisp.dhis.category.CategoryOption;
-import org.hisp.dhis.common.IdScheme;
-import org.hisp.dhis.common.IdentifiableObject;
-import org.hisp.dhis.common.IdentifiableObjectManager;
-import org.hisp.dhis.common.MergeMode;
-import org.hisp.dhis.common.ValueType;
+import org.hisp.dhis.common.*;
```
Comment (mortenoh) [inline dhis-2/dhis-services/dhis-service-dxf2/src/test/java/org/hisp/dhis/dxf2/metadata/objectbundle/ObjectBundleServiceTest.java]: Make sure your import collapse options says 10 or more

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/metadata/objectbundle/hooks/EmbeddedObjectObjectBundleHook.java [hunk 2]
```diff
@@ -53,6 +58,55 @@
...
+    {
+        List<ErrorReport> errors = new ArrayList<>();
+
+        Class<? extends IdentifiableObject> klass = object.getClass();
+        Schema schema = schemaService.getDynamicSchema( klass );
+
+        schema.getEmbeddedObjectProperties().keySet()
+            .stream()
+            .forEach( propertyName -> {
+
```
Comment (mortenoh) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/metadata/objectbundle/hooks/EmbeddedObjectObjectBundleHook.java]: remove newline

File: dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/SystemController.java [hunk 3]
```diff
@@ -237,6 +237,24 @@ public void getTasksExtendedJson( @PathVariable( "jobType" ) String jobType, Htt
         renderService.toJson( response.getOutputStream(), notifications );
     }
 
+    @RequestMapping( value = "/tasks/extended/{jobType}", method = RequestMethod.GET, produces = { "*/*",
```
Comment (mortenoh) [inline dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/SystemController.java]: Was the 2 controller methods meant to be part of this PR?

Comment (enricocolasante) [reply dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/SystemController.java]: Kind of...
The Frontend should call the extended endpoint, but they are not using versioned endpoints, and as it is now, they have no way to do so.
This is why for me the right way is to separate the 2 versioned method.

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/metadata/objectbundle/hooks/EmbeddedObjectObjectBundleHook.java [hunk 4]
```diff
@@ -53,6 +58,54 @@
...
+                    }
+                }
+            } );
+
+        return errors;
+    }
+
+    private List<ErrorReport> formatEmbeddedErrorReport( List<ErrorReport> errors, String embeddedPropertyName )
+    {
+        errors.stream()
```
Comment (mortenoh) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/metadata/objectbundle/hooks/EmbeddedObjectObjectBundleHook.java]: No need to use stream() here? can just do normal for loop

File: dhis-2/dhis-services/dhis-service-dxf2/src/test/java/org/hisp/dhis/dxf2/metadata/objectbundle/ObjectBundleServiceTest.java [hunk 5]
```diff
@@ -1123,6 +1127,24 @@ public void testUpdateMetadataWithValidationRules() throws IOException
...
+            RenderFormat.JSON );
+
+        ObjectBundleParams params = new ObjectBundleParams();
+        params.setObjectBundleMode( ObjectBundleMode.COMMIT );
+        params.setImportStrategy( ImportStrategy.CREATE );
+        params.setObjects( metadata );
+
+        ObjectBundle bundle = objectBundleService.create( params );
+        ObjectBundleValidationReport validate = objectBundleValidationService.validate( bundle );
+        assertFalse( validate.getErrorReports().isEmpty() );
```
Comment (mortenoh) [inline dhis-2/dhis-services/dhis-service-dxf2/src/test/java/org/hisp/dhis/dxf2/metadata/objectbundle/ObjectBundleServiceTest.java]: Could be useful here to check that you are getting the error you are expecting, what if there are other errors that are not related to what you are testing for

File: dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/SystemController.java (hunk 3 above)
Comment (mortenoh) [reply dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/SystemController.java]: Yes, but its not related to the expression fixes.. ?

Comment (enricocolasante) [reply dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/SystemController.java]: Removed
